### PR TITLE
Implement #95

### LINKED ISSUE #95: Closure-convert escaping lambdas and indirect calls
## Scope
Implement closure conversion for ordinary escaping lambdas and closure-valued calls in checked backend programs. Capture free variables into environments, lift closure entry functions, rewrite closure construction, and lower indirect application through the ABI from the representation sub-issue.

## Acceptance Criteria
- Current backend rejection specs for escaping lambdas are converted into LLVM success specs where source checking accepts the program.
- New fixtures cover returned functions, local functions stored in variables, and closure calls across let boundaries.
- Generated LLVM validates with `llvm-as`, and representative cases pass the existing object-code smoke path.
- Direct first-order calls continue to use the existing direct path where possible.

## Dependencies / Blockers
- Depends on `Define the LLVM closure ABI and backend IR representation`.
- Blocks partial application, function-valued ADT fields, and recursive higher-order flow siblings.

## Compatibility With Sibling Sub-Issues
This sub-issue owns general lambda capture and closure calls. It should not implement underapplied-call packaging, constructor field representation, or recursive polymorphic edge cases except as minimal tests for the shared conversion path.

### PR DESCRIPTION
**Pull request by soulomoon**
Closes #95.

## Implementation Plan

---
issue_number: 95
pr_number: 112
branch: codex/issue-95
---

## Implementation Plan

Evidence inspected: issue #95 is open and asks for closure conversion of ordinary escaping lambdas and closure-valued calls; PR #112 is open from `codex/issue-95` to `master`; the branch is clean apart from the watcher start commit. `MLF.Backend.IR` already has `BackendClosure` and `BackendClosureCall`, and `MLF.Backend.LLVM.Lower` already lowers explicit closure IR. The missing work is in the checked-program conversion path, where `MLF.Backend.Convert` still emits ordinary `BackendLam`/`BackendApp` shapes for source lambdas and applications.

1. Reconfirm preconditions in the implementation turn: read the watcher-written issue plan, verify `git status --short --branch`, confirm PR #112 still targets `codex/issue-95`, and do not touch watcher-owned runtime files.

2. Add failing tests first. In `test/BackendConvertSpec.hs`, add source-level checked-program cases for: a function value returned across a let boundary, a local function stored/aliased in another variable, a closure-valued call through one or more lets, and a guard that direct first-order calls still produce `BackendApp`/direct function IR where they are non-escaping. In `test/BackendLLVMSpec.hs` or the ProgramSpec-to-LLVM parity matrix, add source-to-LLVM success coverage that validates with `llvm-as`; mark at least one representative closure case for the existing object-code smoke path. Keep the raw explicit-IR `BackendLam` escaping rejection as a fail-closed guard unless a source-level rejection test is found and converted.

3. Implement closure conversion in `src/MLF/Backend/Convert.hs`. Add a small conversion context/state for closure-valued locals and fresh closure entry names, using deterministic names based on the current binding plus an index and avoiding global/generated collisions. Preserve the leading lambda spine of top-level and otherwise direct first-order function bodies so existing direct calls remain direct. For lambdas in value/escaping positions, collect the leading lambda parameters, compute free lexical term captures from the conversion environment, convert capture expressions and the body with captures/params in scope, and emit `BackendClosure` with the original arrow type.

4. Add let-use escape handling. For local let RHSs that are function values, keep the existing direct local-function shape when every use is a direct first-order call. Closure-convert the RHS and mark the binder closure-valued when it is returned, aliased, stored, captured, or otherwise used as a value. Propagate closure-valued shape through simple let aliases so calls across let boundaries become `BackendClosureCall`.

5. Rewrite closure calls during conversion. When converting an application whose callee is a definite closure value or a closure-valued local, collect the full value-argument spine and emit `BackendClosureCall`; leave global/direct local/immediate lambda call heads as `BackendApp`. Continue to fail closed for out-of-scope partial-application packaging, function-valued ADT field representation requiring captured environments, and recursive higher-order flows unless they are already supported by the existing explicit IR/lowering path.

6. Adjust LLVM lowering only where explicit closure values require it. Keep `lowerBackendType` strict for ordinary arrow types, but add/use a runtime-result helper that maps arrow-valued explicit closure results to `LLVMPtr` for closure-call results and functions that return explicit closures. Do not weaken the raw escaping `BackendLam` rejection.

7. Update durable docs for the changed boundary. In `docs/architecture.md`, replace the statement that full source-to-source closure conversion is future work with the narrower current state: ordinary checked-program escaping lambdas and closure calls are converted to explicit backend closure IR, while partial application, function-valued ADT fields needing captured environments, and recursive higher-order flows remain future/fail-closed. Add a concise `CHANGELOG.md` entry because this is meaningful backend progress.

8. Validate in increasing scope: run focused BackendConvert tests for the new conversion cases, focused BackendLLVM/parity tests for closure LLVM emission, then `cabal build all && cabal test` before reporting implementation complete. If LLVM tools are missing, record which `llvm-as`/object-code validations were skipped by the test harness.

9. Publish in the implementation turn only after validation: run `gh auth status` and `gh auth setup-git`, inspect `git status --short` and relevant diffs, stage only issue-related source/test/docs files, commit as the prepared `codex-watcher` identity with an imperative message, and push `codex/issue-95` to update PR #112.

---
Implementation plan synced by codex-watcher before implementation starts. Implementation commits will be pushed to this PR.

### CODE REVIEW

File: src/MLF/Backend/Convert.hs [hunk 1]
```diff
@@ -1743,49 +1822,86 @@ convertOrdinaryTerm context env term resultTy0 =
...
+  Env ->
+  ClosureScope ->
+  BackendType ->
+  ElabTerm ->
+  ConvertM BackendExpr
+convertApplicationTerm context env scope resultTy term =
+  case collectApps term of
+    (headTerm, args)
+      | not (null args),
+        isClosureHeadTerm scope headTerm -> do
```
Comment (soulomoon) [inline src/MLF/Backend/Convert.hs]: This only recognizes closure calls when the callee is a closure-valued local, so closure-valued top-level bindings still fall through to the direct `BackendApp`/global-call path. For example, the frontend accepts:

```mlfp
def maker : Int -> Int = let captured : Int = 41 in \(x : Int) captured;
def main : Int = maker 0;
```

but `emit-backend` fails with `Backend LLVM arity mismatch for Main__maker: expected 0, got 1` because `maker` was lowered as a zero-argument function returning a closure pointer. This is still a closure-valued call and needs to be rewritten to the closure-call path or rejected before LLVM lowering.

File: src/MLF/Backend/Convert.hs [hunk 2]
```diff
@@ -1802,11 +1918,231 @@ applicationResultType resultTy funExpr =
...
+          go underLambda body
+        EUnroll body ->
+          go underLambda body
+
+convertLambdaClosure :: LambdaMode -> ConvertContext -> Env -> ClosureScope -> BackendType -> ElabTerm -> ConvertM BackendExpr
+convertLambdaClosure mode context env scope resultTy term = do
+  let (rawParams, body) = collectClosureLams term
+  when (null rawParams) $
+    liftEitherConvert (Left (BackendUnsupportedCaseShape "closure conversion expected a lambda"))
+  (params, bodyExpected) <- closureBackendParams context resultTy rawParams
```
Comment (soulomoon) [inline src/MLF/Backend/Convert.hs:2192]: `closureBackendParams` can accept fewer collected lambda parameters than there are arrows in `resultTy`, which lets conversion emit an invalid `BackendClosure` when an escaping lambda returns another lambda behind a let. Repro:

```mlfp
def main : Int -> Int -> Int =
  let captured : Int = 41 in
  let f : Int -> Int -> Int = \(x : Int) let y : Int = captured in \(z : Int) y in
  f;
```

The frontend accepts this, but backend conversion fails validation with `BackendClosureParameterArityMismatch ... 1 2`. Either the returned closure shape needs to be represented explicitly, or this source shape should fail closed before invalid IR is built.

**Review by soulomoon — COMMENTED** [review]
I found two closure-conversion gaps that are accepted by the frontend but still fail in the backend.

**Comment by soulomoon** [comment]
Review findings require rework at commit `038b96f4a904a2039ff93fa91c5f9d512e123104`.

Findings to address:
- Inline comment added on src/MLF/Backend/Convert.hs:1850: closure-valued top-level bindings are not recognized as closure call heads, so an accepted program like `maker : Int -> Int = let captured = 41 in \x -> captured; main = maker 0` still reaches LLVM as a direct global call and fails with `Backend LLVM arity mismatch`.
- Inline comment added on src/MLF/Backend/Convert.hs:2040: escaping lambdas that return another lambda behind a let can emit invalid `BackendClosure` IR with too few closure parameters; an accepted returned-function program fails validation with `BackendClosureParameterArityMismatch`.

File: src/MLF/Backend/Convert.hs [hunk 3]
```diff
@@ -1618,87 +1675,109 @@ candidateDataResultTypes context ty =
...
               _ -> (normalizeBackendTypeForContext context rawParamBackendTy, Nothing)
-      let paramEnvTy =
+          paramEnvTy =
             case backendTypeToElabType paramBackendTy of
               Just canonicalTy -> canonicalTy
               Nothing -> paramTy
-      bodyExpr <- convertTermExpected context (extendTermEnv name paramEnvTy env) bodyExpected body
-      Right
+          bodyMode = directLambdaBodyMode bodyExpected body
+          bodyScope = extendClosureScopeTerm name paramEnvTy False scope
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply src/MLF/Backend/Convert.hs]: This still applies at `dfe9d1d9575fe403f77d784893bb492314bb7f62`. Repro:

```mlfp
def use : (Int -> Int) -> Int = \(f : Int -> Int) f 1;
def main : Int =
  let captured : Int = 41 in
  let f : Int -> Int = \(x : Int) captured in
  use f;
```

The frontend returns `41`, but `emit-backend` still emits `call i64 %"__llvm.malloc.2"(i64 1)`, calling the closure object pointer directly instead of loading the closure code/env and using `BackendClosureCall`.

**Comment by soulomoon** [comment]
Review findings require rework at commit `dfe9d1d9575fe403f77d784893bb492314bb7f62`.

Findings to address:
- Unresolved review thread: This still applies at `dfe9d1d9575fe403f77d784893bb492314bb7f62`. Repro:

```mlfp
def use : (Int -> Int) -> Int = \(f : Int -> Int) f 1;
def main : Int =
  let captured : Int = 41 in
  let f : Int -> Int = \(x : Int) captured in
  use f;
```

The frontend returns `41`, but `emit-backend` still emits `call i64 %"__llvm.malloc.2"(i64 1)`, calling the closure object pointer directly instead of loading the closure code/env and using `BackendClosureCall`.

File: src/MLF/Backend/Convert.hs [hunk 4]
```diff
@@ -2015,24 +2441,26 @@ convertConstructorApplication context env term resultTy =
...
+                  ( "constructor result type does not match expected result for `"
+                      ++ backendConstructorName constructor
+                      ++ "`"
+                  )
               )
           )
-      argExprs <- zipWithM (convertTermExpected context env . Just) fields args
-      mapM_ (checkConstructorArgumentType context ownerContext typeBounds constructor) (zip [0 :: Int ..] (zip fields argExprs))
-      Right
+      argExprs <- zipWithM (convertTermExpectedMode DirectLambda context env scope . Just) fields args
```
Comment (soulomoon) [inline src/MLF/Backend/Convert.hs:2552]: This now allows a closure-valued local to be converted as an argument to a function-typed constructor field, but case binders for those fields are still lowered as ordinary function pointers. For example:

```mlfp
module Main export (FnBox(..), main) {
  data FnBox = FnBox : (Int -> Int) -> FnBox;
  def main : Int =
    let captured : Int = 41 in
    let f : Int -> Int = \(x : Int) captured in
    case FnBox f of { FnBox g -> g 0 };
}
```

`emit-backend` succeeds and emits `call i64 %"__llvm.malloc.2"(i64 0)`, calling the closure object pointer as if it were raw code. Since the PR plan keeps function-valued ADT fields needing captured environments future/fail-closed, conversion should reject closure-valued arguments for these fields, or the field/case path needs to use the closure ABI end-to-end.

**Review by soulomoon — COMMENTED** [review]
I found one closure-conversion gap that should stay fail-closed until function-valued ADT fields get a closure-aware representation.

File: src/MLF/Backend/Convert.hs [hunk 5]
```diff
@@ -1802,11 +1959,273 @@ applicationResultType resultTy funExpr =
...
+backendExprIsClosureValue context scope =
+  \case
+    BackendClosure {} -> True
+    BackendVar _ name -> Set.member name (closureScopeLocals scope) || Set.member name (ccClosureGlobals context)
+    BackendTyApp _ fun _ -> backendExprIsClosureValue context scope fun
+    BackendLet _ name _ rhs body ->
+      let bodyScope =
+            if backendExprIsClosureValue context scope rhs
+              then scope {closureScopeLocals = Set.insert name (closureScopeLocals scope)}
+              else scope
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply src/MLF/Backend/Convert.hs]: Still applies at `2bf0aafb60fc43c20a95ecc50d7be267dc151d76`: the `BackendLet` false branch in `backendExprIsClosureValue` still leaves `closureScopeLocals` unchanged instead of deleting the rebound name, so shadowed closure-typed names can retain an outer closure classification.

File: src/MLF/Backend/Convert.hs (hunk 4 above)
Comment (soulomoon) [reply src/MLF/Backend/Convert.hs:2552]: Still applies at `2bf0aafb60fc43c20a95ecc50d7be267dc151d76`: the source-level `FnBox` repro still emits `call i64 %"__llvm.malloc.2"(i64 0)`, calling the closure object pointer as a raw code pointer instead of failing closed or using the closure ABI for the function-valued ADT field path.

**Comment by soulomoon** [comment]
Review findings require rework at commit `2bf0aafb60fc43c20a95ecc50d7be267dc151d76`.

Findings to address:
- Unresolved review thread: Still applies at `2bf0aafb60fc43c20a95ecc50d7be267dc151d76`: the `BackendLet` false branch in `backendExprIsClosureValue` still leaves `closureScopeLocals` unchanged instead of deleting the rebound name, so shadowed closure-typed names can retain an outer closure classification.
- Unresolved review thread: Still applies at `2bf0aafb60fc43c20a95ecc50d7be267dc151d76`: the source-level `FnBox` repro still emits `call i64 %"__llvm.malloc.2"(i64 0)`, calling the closure object pointer as a raw code pointer instead of failing closed or using the closure ABI for the function-valued ADT field path.

File: src/MLF/Backend/Convert.hs [hunk 6]
```diff
@@ -1802,11 +1959,279 @@ applicationResultType resultTy funExpr =
...
+                      then Set.insert
+                      else Set.delete
+                  )
+                    name
+                    (closureScopeLocals scope)
+              }
+       in backendExprIsClosureValue context bodyScope body
+    BackendCase {backendAlternatives = alternatives} ->
+      all (backendExprIsClosureValue context scope . backendAltBody) (NE.toList alternatives)
+    _ -> False
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply src/MLF/Backend/Convert.hs:2101]: Still applies at `9aa989ba277ac3b8759023e898d914841a3595bd`. I reproduced this with `def maker : forall a. a -> Int = let captured : Int = 41 in \(x : a) captured; def main : Int = maker 0;`; `emit-backend` fails with `Backend LLVM arity mismatch for Main__maker: expected 0, got 1`. `backendExprIsClosureValue` still handles `BackendTyApp` but not `BackendTyAbs`, so the closure-global fixed point misses type-abstracted globals whose body produces a closure.

File: src/MLF/Backend/IR.hs [hunk 7]
```diff
@@ -860,6 +900,123 @@ firstClosureValueName (Nothing : rest) =
...
+          | otherwise -> go aliases body
+        BackendApp _ fun arg ->
+          go aliases fun || go aliases arg
+        BackendLet _ name _ rhs body
+          | Set.member name aliases -> go aliases rhs
+          | otherwise ->
+              let aliasesForBody =
+                    if closureCallHeadReferencesAny aliases rhs
+                      then Set.insert name aliases
+                      else aliases
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply src/MLF/Backend/IR.hs:943]: Still applies at `9aa989ba277ac3b8759023e898d914841a3595bd`. I reproduced the nested alias case with `let g : Int -> Int = (let h : Int -> Int = f in h) in g 1`; `emit-backend` fails validation with `BackendClosureCallExpectedClosureValue`. The alias analysis still only promotes a let binder when the RHS head directly references the alias set, so it misses nested let forwarding.

**Comment by soulomoon** [comment]
Review findings require rework at commit `9aa989ba277ac3b8759023e898d914841a3595bd`.

Findings to address:
- Unresolved review thread: Still applies at `9aa989ba277ac3b8759023e898d914841a3595bd`. I reproduced this with `def maker : forall a. a -> Int = let captured : Int = 41 in \(x : a) captured; def main : Int = maker 0;`; `emit-backend` fails with `Backend LLVM arity mismatch for Main__maker: expected 0, got 1`. `backendExprIsClosureValue` still handles `BackendTyApp` but not `BackendTyAbs`, so the closure-global fixed point misses type-abstracted globals whose body produces a closure.
- Unresolved review thread: Still applies at `9aa989ba277ac3b8759023e898d914841a3595bd`. I reproduced the nested alias case with `let g : Int -> Int = (let h : Int -> Int = f in h) in g 1`; `emit-backend` fails validation with `BackendClosureCallExpectedClosureValue`. The alias analysis still only promotes a let binder when the RHS head directly references the alias set, so it misses nested let forwarding.

File: src/MLF/Backend/Convert.hs [hunk 8]
```diff
@@ -1858,11 +2018,280 @@ applicationResultType resultTy funExpr =
...
+                  ( if backendExprIsClosureValue context scope rhs
+                      then Set.insert
+                      else Set.delete
+                  )
+                    name
+                    (closureScopeLocals scope)
+              }
+       in backendExprIsClosureValue context bodyScope body
+    BackendCase {backendAlternatives = alternatives} ->
+      all (backendExprIsClosureValue context scope . backendAltBody) (NE.toList alternatives)
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply src/MLF/Backend/Convert.hs]: Still applies at `1cb0bd0198c2f6ec0ea53cea352561ffc55b1f46`: the `BackendCase` branch still evaluates every alternative body under the unchanged `ClosureScope`, so constructor pattern binders do not shadow names in `closureScopeLocals`. A function-typed pattern binder that reuses an outer closure-local name can still be misclassified as a closure value and routed to the closure-call path incorrectly.

**Comment by soulomoon** [comment]
Review findings require rework at commit `1cb0bd0198c2f6ec0ea53cea352561ffc55b1f46`.

Findings to address:
- Unresolved review thread: Still applies at `1cb0bd0198c2f6ec0ea53cea352561ffc55b1f46`: the `BackendCase` branch still evaluates every alternative body under the unchanged `ClosureScope`, so constructor pattern binders do not shadow names in `closureScopeLocals`. A function-typed pattern binder that reuses an outer closure-local name can still be misclassified as a closure value and routed to the closure-call path incorrectly.

**Review by soulomoon — COMMENTED** [review]
LGTM at 8b1dedaa6794daa40313b68e40a902d8929bdc95. The PR satisfies the plan and linked issue scope, remains mergeable, and validation passed with `cabal build all && cabal test`.

Reviewed commit: `8b1dedaa6794daa40313b68e40a902d8929bdc95`

Submitted by Haskell PR review watcher as a clean review comment before merge.
